x-pack/plugins/apm/public/components/app/Correlations/useFieldNames.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { memoize, take } from 'lodash';
+import { useEffect, useMemo, useState } from 'react';
+import { ProcessorEvent } from '../../../../common/processor_event';
+import { isRumAgentName } from '../../../../common/agent_name';
+import { useApmServiceContext } from '../../../context/apm_service/use_apm_service_context';
+import { useDynamicIndexPatternFetcher } from '../../../hooks/use_dynamic_index_pattern';
+
+interface IndexPattern {
+  fields: Array<{ name: string }>;
+}
+
+export function useFieldNames() {
+  const { agentName } = useApmServiceContext();
+  const isRumAgent = isRumAgentName(agentName);
+  const { indexPattern } = useDynamicIndexPatternFetcher(
+    ProcessorEvent.transaction
+  );
+
+  const [defaultFieldNames, setDefaultFieldNames] = useState(
+    getDefaultFieldNames(indexPattern, isRumAgent)
+  );
```

Comment:
It feels off that this is stateful. If for perf reasons you can use `useMemo`. 
But first try to strip it down and see if it's needed

Reply:
It's stateful because `useDynamicIndexPatternFetcher` is an async fetcher, so it needs to trigger a re-render when the response comes back.